provider/fastly Allows for conditional settings across fastly

travertischio opened the pull request:
In Fastly:
* Allows user to set `cache_condition` for Gzip
* Allows user to set `cache_condition`, `request_condition` and `response_condition` for Header
* Adds optional `response_condition` setting for S3Logging
* Adds optional `response_condition` setting for Papertrail

- [x] Tests included
- [x] Website documentation included

#7554

stack72 commented:
Hi @travertischio 

Thanks for the work here - unfortunately, this is causing some of the tests to start failing:

```
% make testacc TEST=./builtin/providers/fastly                                                                       ✭
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/02/10 11:30:35 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/fastly -v  -timeout 120m
=== RUN   TestAccFastlyIPRanges
--- PASS: TestAccFastlyIPRanges (1.70s)
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccFastlyServiceV1CacheSetting_basic
--- PASS: TestAccFastlyServiceV1CacheSetting_basic (40.54s)
=== RUN   TestAccFastlyServiceV1_conditional_basic
--- PASS: TestAccFastlyServiceV1_conditional_basic (15.52s)
=== RUN   TestFastlyServiceV1_FlattenGzips
--- PASS: TestFastlyServiceV1_FlattenGzips (0.00s)
=== RUN   TestAccFastlyServiceV1_gzips_basic
--- FAIL: TestAccFastlyServiceV1_gzips_basic (3.86s)
	testing.go:265: Step 0 error: Error applying: 1 error(s) occurred:

		* fastly_service_v1.foo: 1 error(s) occurred:

		* fastly_service_v1.foo: 400 - Bad Request
		Message: An error occurred while connecting to the fastly API, please try your request again.
		Detail: No condition named '!beresp.cachable'
=== RUN   TestFastlyServiceV1_BuildHeaders
--- PASS: TestFastlyServiceV1_BuildHeaders (0.00s)
=== RUN   TestAccFastlyServiceV1_headers_basic
--- FAIL: TestAccFastlyServiceV1_headers_basic (28.36s)
	testing.go:265: Step 1 error: Error applying: 1 error(s) occurred:

		* fastly_service_v1.foo: 1 error(s) occurred:

		* fastly_service_v1.foo: 400 - Bad Request
		Message: An error occurred while connecting to the fastly API, please try your request again.
		Detail: No condition named 'req.url ~ "^/foo/bar$"'
=== RUN   TestAccFastlyServiceV1_healthcheck_basic
--- PASS: TestAccFastlyServiceV1_healthcheck_basic (36.47s)
=== RUN   TestAccFastlyServiceV1_papertrail_basic
--- FAIL: TestAccFastlyServiceV1_papertrail_basic (3.79s)
	testing.go:265: Step 0 error: Error applying: 1 error(s) occurred:

		* fastly_service_v1.foo: 1 error(s) occurred:

		* fastly_service_v1.foo: 400 - Bad Request
		Message: An error occurred while connecting to the fastly API, please try your request again.
		Detail: No condition named 'resp.status >= 400 && resp.status < 600'
=== RUN   TestAccFastlyServiceV1RequestSetting_basic
--- PASS: TestAccFastlyServiceV1RequestSetting_basic (17.47s)
=== RUN   TestAccFastlyServiceV1_s3logging_basic
--- FAIL: TestAccFastlyServiceV1_s3logging_basic (4.60s)
	testing.go:265: Step 0 error: Error applying: 1 error(s) occurred:

		* fastly_service_v1.foo: 1 error(s) occurred:

		* fastly_service_v1.foo: 400 - Bad Request
		Message: An error occurred while connecting to the fastly API, please try your request again.
		Detail: No condition named 'resp.status == 418'
=== RUN   TestAccFastlyServiceV1_s3logging_s3_env
--- PASS: TestAccFastlyServiceV1_s3logging_s3_env (20.32s)
=== RUN   TestAccFastlyServiceV1_s3logging_formatVersion
--- PASS: TestAccFastlyServiceV1_s3logging_formatVersion (20.15s)
=== RUN   TestResourceFastlyFlattenDomains
--- PASS: TestResourceFastlyFlattenDomains (0.00s)
=== RUN   TestResourceFastlyFlattenBackend
--- PASS: TestResourceFastlyFlattenBackend (0.00s)
=== RUN   TestAccFastlyServiceV1_updateDomain
--- PASS: TestAccFastlyServiceV1_updateDomain (38.67s)
=== RUN   TestAccFastlyServiceV1_updateBackend
--- PASS: TestAccFastlyServiceV1_updateBackend (39.42s)
=== RUN   TestAccFastlyServiceV1_basic
--- PASS: TestAccFastlyServiceV1_basic (16.09s)
=== RUN   TestAccFastlyServiceV1_disappears
--- PASS: TestAccFastlyServiceV1_disappears (7.36s)
=== RUN   TestAccFastlyServiceV1_VCL_basic
--- PASS: TestAccFastlyServiceV1_VCL_basic (38.67s)
=== RUN   TestValidateS3FormatVersion
--- PASS: TestValidateS3FormatVersion (0.00s)
FAIL
exit status 1
FAIL	github.com/hashicorp/terraform/builtin/providers/fastly	333.014s
make: *** [testacc] Error 1
```

If you could, could you have a look? 

Thanks

Paul

travertischio commented:
@stack72 I believed I have fixed all the issues here.

stack72 commented:
Hi @travertischio 

Still getting the same issues it seems:

```
% make testacc TEST=./builtin/providers/fastly                                                                     ✹ ✭
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/02/13 11:54:54 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/fastly -v  -timeout 120m
=== RUN   TestAccFastlyIPRanges
--- PASS: TestAccFastlyIPRanges (1.67s)
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccFastlyServiceV1CacheSetting_basic
--- PASS: TestAccFastlyServiceV1CacheSetting_basic (42.76s)
=== RUN   TestAccFastlyServiceV1_conditional_basic
--- PASS: TestAccFastlyServiceV1_conditional_basic (17.01s)
=== RUN   TestFastlyServiceV1_FlattenGzips
--- PASS: TestFastlyServiceV1_FlattenGzips (0.00s)
=== RUN   TestAccFastlyServiceV1_gzips_basic
--- FAIL: TestAccFastlyServiceV1_gzips_basic (5.72s)
	testing.go:265: Step 0 error: Error applying: 1 error(s) occurred:

		* fastly_service_v1.foo: 1 error(s) occurred:

		* fastly_service_v1.foo: [ERR] Invalid configuration for Fastly Service (2xMMROyY1kdHonHwPTuO0M): Unknown variable 'beresp.cachable'
		At: (input Line 116 Pos 8)
		 if ( !beresp.cachable ) {
		-------###############----
		Running VCC-compiler failed
=== RUN   TestFastlyServiceV1_BuildHeaders
--- PASS: TestFastlyServiceV1_BuildHeaders (0.00s)
=== RUN   TestAccFastlyServiceV1_headers_basic
--- FAIL: TestAccFastlyServiceV1_headers_basic (38.02s)
	testing.go:265: Step 1 error: Check failed: Check 5/8 error: fastly_service_v1.foo: Attribute 'header.1147514417.source' not found
=== RUN   TestAccFastlyServiceV1_healthcheck_basic
--- PASS: TestAccFastlyServiceV1_healthcheck_basic (40.62s)
=== RUN   TestAccFastlyServiceV1_papertrail_basic
--- PASS: TestAccFastlyServiceV1_papertrail_basic (38.33s)
=== RUN   TestAccFastlyServiceV1RequestSetting_basic
--- PASS: TestAccFastlyServiceV1RequestSetting_basic (18.07s)
=== RUN   TestAccFastlyServiceV1_s3logging_basic
--- FAIL: TestAccFastlyServiceV1_s3logging_basic (4.94s)
	testing.go:265: Step 0 error: Error applying: 1 error(s) occurred:

		* fastly_service_v1.foo: 1 error(s) occurred:

		* fastly_service_v1.foo: [ERR] Invalid configuration for Fastly Service (5VQ8BeHxDd3m0ZfieIo35A): Invalid condition '=' on numeric variable
		  only '==', '!=', '<', '>', '<=' and '>=' are legal
		(input Line 285 Pos 19)
		  if( resp.status = 418 ) {
		------------------#--------
		Running VCC-compiler failed
=== RUN   TestAccFastlyServiceV1_s3logging_s3_env
--- PASS: TestAccFastlyServiceV1_s3logging_s3_env (16.38s)
=== RUN   TestAccFastlyServiceV1_s3logging_formatVersion
--- PASS: TestAccFastlyServiceV1_s3logging_formatVersion (18.16s)
=== RUN   TestResourceFastlyFlattenDomains
--- PASS: TestResourceFastlyFlattenDomains (0.00s)
=== RUN   TestResourceFastlyFlattenBackend
--- PASS: TestResourceFastlyFlattenBackend (0.00s)
=== RUN   TestAccFastlyServiceV1_updateDomain
--- PASS: TestAccFastlyServiceV1_updateDomain (39.14s)
=== RUN   TestAccFastlyServiceV1_updateBackend
--- PASS: TestAccFastlyServiceV1_updateBackend (40.35s)
=== RUN   TestAccFastlyServiceV1_basic
--- PASS: TestAccFastlyServiceV1_basic (16.67s)
=== RUN   TestAccFastlyServiceV1_disappears
--- PASS: TestAccFastlyServiceV1_disappears (8.99s)
=== RUN   TestAccFastlyServiceV1_VCL_basic
--- PASS: TestAccFastlyServiceV1_VCL_basic (37.07s)
=== RUN   TestValidateS3FormatVersion
--- PASS: TestValidateS3FormatVersion (0.00s)
FAIL
exit status 1
FAIL	github.com/hashicorp/terraform/builtin/providers/fastly	383.940s
make: *** [testacc] Error 1
```

Also I had to `make fmt` the branch:

```
diff --git a/builtin/providers/fastly/resource_fastly_service_v1_papertrail_test.go b/builtin/providers/fastly/resource_fastly_service_v1_papertrail_test.go
index 4d1f6c8..69500f2 100644
--- a/builtin/providers/fastly/resource_fastly_service_v1_papertrail_test.go
+++ b/builtin/providers/fastly/resource_fastly_service_v1_papertrail_test.go
@@ -26,11 +26,11 @@ func TestAccFastlyServiceV1_papertrail_basic(t *testing.T) {
        }

        log2 := gofastly.Papertrail{
-               Version:           "1",
-               Name:              "papertrailtesting2",
-               Address:           "test2.papertrailapp.com",
-               Port:              uint(8080),
-               Format:            "%h %l %u %t %r %>s",
+               Version: "1",
+               Name:    "papertrailtesting2",
+               Address: "test2.papertrailapp.com",
+               Port:    uint(8080),
+               Format:  "%h %l %u %t %r %>s",
        }
```

travertischio commented:
@stack72 Sorry about the formatting issue in the previous commits. I adjusted the testing I added to the syntax issues that the errors you copied here were showing and moved around some testing so that whole schemas could be checked en mass like the other services in fastly. How is this now?

stack72 commented:
Still not working i'm afraid @travertischio 

```
% make testacc TEST=./builtin/providers/fastly                                                                     1 ↵
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/02/13 17:39:36 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/fastly -v  -timeout 120m
=== RUN   TestAccFastlyIPRanges
--- PASS: TestAccFastlyIPRanges (1.70s)
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccFastlyServiceV1CacheSetting_basic
--- PASS: TestAccFastlyServiceV1CacheSetting_basic (40.46s)
=== RUN   TestAccFastlyServiceV1_conditional_basic
--- PASS: TestAccFastlyServiceV1_conditional_basic (18.84s)
=== RUN   TestFastlyServiceV1_FlattenGzips
--- PASS: TestFastlyServiceV1_FlattenGzips (0.00s)
=== RUN   TestAccFastlyServiceV1_gzips_basic
--- FAIL: TestAccFastlyServiceV1_gzips_basic (0.00s)
	testing.go:265: Step 0 error: Error loading configuration: Error parsing /var/folders/gd/yv86v7pd1cd5dpzjrw8702k00000gn/T/tf-test230354721/main.tf: At 19:29: illegal char
=== RUN   TestFastlyServiceV1_BuildHeaders
--- PASS: TestFastlyServiceV1_BuildHeaders (0.00s)
=== RUN   TestAccFastlyServiceV1_headers_basic
--- FAIL: TestAccFastlyServiceV1_headers_basic (9.43s)
	testing.go:265: Step 0 error: Check failed: Check 2/4 error: Bad match Header match, expected (&fastly.Header{ServiceID:"4EAmEaDqNXqS8CPAiaryga", Version:"1", Name:"remove x-amz-request-id", Action:"delete", IgnoreIfSet:false, Type:"cache", Destination:"http.x-amz-request-id", Source:"", Regex:"", Substitution:"", Priority:0x0, RequestCondition:"", CacheCondition:"", ResponseCondition:""}), got (&fastly.Header{ServiceID:"4EAmEaDqNXqS8CPAiaryga", Version:"1", Name:"remove x-amz-request-id", Action:"delete", IgnoreIfSet:false, Type:"cache", Destination:"http.x-amz-request-id", Source:"", Regex:"", Substitution:"", Priority:0x64, RequestCondition:"", CacheCondition:"", ResponseCondition:""})
	testing.go:329: Error destroying resource! WARNING: Dangling resources
		may exist. The full state and error is shown below.

		Error: Error loading configuration: Error parsing /var/folders/gd/yv86v7pd1cd5dpzjrw8702k00000gn/T/tf-test929202171/main.tf: At 40:29: illegal char

		State: fastly_service_v1.foo:
		  ID = 4EAmEaDqNXqS8CPAiaryga
		  active_version = 1
		  backend.# = 1
		  backend.510555480.address = aws.amazon.com
		  backend.510555480.auto_loadbalance = true
		  backend.510555480.between_bytes_timeout = 10000
		  backend.510555480.connect_timeout = 1000
		  backend.510555480.error_threshold = 0
		  backend.510555480.first_byte_timeout = 15000
		  backend.510555480.max_conn = 200
		  backend.510555480.name = amazon docs
		  backend.510555480.port = 80
		  backend.510555480.request_condition =
		  backend.510555480.shield =
		  backend.510555480.ssl_check_cert = true
		  backend.510555480.ssl_hostname =
		  backend.510555480.weight = 100
		  cache_setting.# = 0
		  condition.# = 0
		  default_host =
		  default_ttl = 3600
		  domain.# = 1
		  domain.3405649020.comment = tf-testing-domain
		  domain.3405649020.name = i9ttqilbhy.notadomain.com
		  force_destroy = true
		  gzip.# = 0
		  header.# = 2
		  header.1374723720.action = delete
		  header.1374723720.cache_condition =
		  header.1374723720.destination = http.x-amz-request-id
		  header.1374723720.ignore_if_set = false
		  header.1374723720.name = remove x-amz-request-id
		  header.1374723720.priority = 100
		  header.1374723720.regex =
		  header.1374723720.request_condition =
		  header.1374723720.response_condition =
		  header.1374723720.source =
		  header.1374723720.substitution =
		  header.1374723720.type = cache
		  header.3124586293.action = delete
		  header.3124586293.cache_condition =
		  header.3124586293.destination = http.Server
		  header.3124586293.ignore_if_set = true
		  header.3124586293.name = remove s3 server
		  header.3124586293.priority = 100
		  header.3124586293.regex =
		  header.3124586293.request_condition =
		  header.3124586293.response_condition =
		  header.3124586293.source =
		  header.3124586293.substitution =
		  header.3124586293.type = cache
		  healthcheck.# = 0
		  name = tf-test-q1newd1rlu
		  papertrail.# = 0
		  request_setting.# = 0
		  s3logging.# = 0
		  vcl.# = 0
=== RUN   TestAccFastlyServiceV1_healthcheck_basic
--- PASS: TestAccFastlyServiceV1_healthcheck_basic (42.74s)
=== RUN   TestAccFastlyServiceV1_papertrail_basic
--- PASS: TestAccFastlyServiceV1_papertrail_basic (38.69s)
=== RUN   TestAccFastlyServiceV1RequestSetting_basic
--- PASS: TestAccFastlyServiceV1RequestSetting_basic (18.32s)
=== RUN   TestAccFastlyServiceV1_s3logging_basic
--- FAIL: TestAccFastlyServiceV1_s3logging_basic (39.23s)
	testing.go:265: Step 1 error: Check failed: Check 2/4 error: Bad match S3 logging match, expected (&fastly.S3{ServiceID:"6DFKyJmBM7EcGcvQDmLzmC", Version:"2", Name:"somebucketlog", BucketName:"fastlytestlogging", Domain:"s3-us-west-2.amazonaws.com", AccessKey:"somekey", SecretKey:"somesecret", Path:"", Period:0xe10, GzipLevel:0x0, Format:"%h %l %u %t %r %>s", FormatVersion:0x1, ResponseCondition:"response_condition_test", TimestampFormat:"%Y-%m-%dT%H:%M:%S.000", Redundancy:"", CreatedAt:(*time.Time)(nil), UpdatedAt:(*time.Time)(nil), DeletedAt:(*time.Time)(nil)}), got (&fastly.S3{ServiceID:"6DFKyJmBM7EcGcvQDmLzmC", Version:"2", Name:"somebucketlog", BucketName:"fastlytestlogging", Domain:"s3-us-west-2.amazonaws.com", AccessKey:"somekey", SecretKey:"somesecret", Path:"", Period:0xe10, GzipLevel:0x0, Format:"%h %l %u %t %r %>s", FormatVersion:0x1, ResponseCondition:"", TimestampFormat:"%Y-%m-%dT%H:%M:%S.000", Redundancy:"", CreatedAt:(*time.Time)(nil), UpdatedAt:(*time.Time)(nil), DeletedAt:(*time.Time)(nil)})
=== RUN   TestAccFastlyServiceV1_s3logging_s3_env
--- PASS: TestAccFastlyServiceV1_s3logging_s3_env (16.67s)
=== RUN   TestAccFastlyServiceV1_s3logging_formatVersion
--- PASS: TestAccFastlyServiceV1_s3logging_formatVersion (15.87s)
=== RUN   TestResourceFastlyFlattenDomains
--- PASS: TestResourceFastlyFlattenDomains (0.00s)
=== RUN   TestResourceFastlyFlattenBackend
--- PASS: TestResourceFastlyFlattenBackend (0.00s)
=== RUN   TestAccFastlyServiceV1_updateDomain
--- PASS: TestAccFastlyServiceV1_updateDomain (38.33s)
=== RUN   TestAccFastlyServiceV1_updateBackend
--- PASS: TestAccFastlyServiceV1_updateBackend (44.39s)
=== RUN   TestAccFastlyServiceV1_basic
--- PASS: TestAccFastlyServiceV1_basic (16.02s)
=== RUN   TestAccFastlyServiceV1_disappears
--- PASS: TestAccFastlyServiceV1_disappears (9.33s)
=== RUN   TestAccFastlyServiceV1_VCL_basic
--- PASS: TestAccFastlyServiceV1_VCL_basic (40.93s)
=== RUN   TestValidateS3FormatVersion
--- PASS: TestValidateS3FormatVersion (0.00s)
FAIL
exit status 1
FAIL	github.com/hashicorp/terraform/builtin/providers/fastly	390.982s
make: *** [testacc] Error 1
```

the errors have changed though


P.

travertischio commented:
Hey @stack72 I am fairly certain that I finally pulled everything together. Thank you bearing with me here.

stack72 commented:
Amazing - thanks for getting this taken care of :) All looking good now!

```
% make testacc TEST=./builtin/providers/fastly
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/02/15 09:45:36 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/fastly -v  -timeout 120m
=== RUN   TestAccFastlyIPRanges
--- PASS: TestAccFastlyIPRanges (1.51s)
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccFastlyServiceV1CacheSetting_basic
--- PASS: TestAccFastlyServiceV1CacheSetting_basic (41.78s)
=== RUN   TestAccFastlyServiceV1_conditional_basic
--- PASS: TestAccFastlyServiceV1_conditional_basic (18.95s)
=== RUN   TestFastlyServiceV1_FlattenGzips
--- PASS: TestFastlyServiceV1_FlattenGzips (0.00s)
=== RUN   TestAccFastlyServiceV1_gzips_basic
--- PASS: TestAccFastlyServiceV1_gzips_basic (45.55s)
=== RUN   TestFastlyServiceV1_BuildHeaders
--- PASS: TestFastlyServiceV1_BuildHeaders (0.00s)
=== RUN   TestAccFastlyServiceV1_headers_basic
--- PASS: TestAccFastlyServiceV1_headers_basic (43.08s)
=== RUN   TestAccFastlyServiceV1_healthcheck_basic
--- PASS: TestAccFastlyServiceV1_healthcheck_basic (40.27s)
=== RUN   TestAccFastlyServiceV1_papertrail_basic
--- PASS: TestAccFastlyServiceV1_papertrail_basic (39.79s)
=== RUN   TestAccFastlyServiceV1RequestSetting_basic
--- PASS: TestAccFastlyServiceV1RequestSetting_basic (18.69s)
=== RUN   TestAccFastlyServiceV1_s3logging_basic
--- PASS: TestAccFastlyServiceV1_s3logging_basic (41.82s)
=== RUN   TestAccFastlyServiceV1_s3logging_s3_env
--- PASS: TestAccFastlyServiceV1_s3logging_s3_env (16.93s)
=== RUN   TestAccFastlyServiceV1_s3logging_formatVersion
--- PASS: TestAccFastlyServiceV1_s3logging_formatVersion (17.67s)
=== RUN   TestResourceFastlyFlattenDomains
--- PASS: TestResourceFastlyFlattenDomains (0.00s)
=== RUN   TestResourceFastlyFlattenBackend
--- PASS: TestResourceFastlyFlattenBackend (0.00s)
=== RUN   TestAccFastlyServiceV1_updateDomain
--- PASS: TestAccFastlyServiceV1_updateDomain (39.32s)
=== RUN   TestAccFastlyServiceV1_updateBackend
--- PASS: TestAccFastlyServiceV1_updateBackend (43.17s)
=== RUN   TestAccFastlyServiceV1_basic
--- PASS: TestAccFastlyServiceV1_basic (18.83s)
=== RUN   TestAccFastlyServiceV1_disappears
--- PASS: TestAccFastlyServiceV1_disappears (10.37s)
=== RUN   TestAccFastlyServiceV1_VCL_basic
--- PASS: TestAccFastlyServiceV1_VCL_basic (42.41s)
=== RUN   TestValidateS3FormatVersion
--- PASS: TestValidateS3FormatVersion (0.00s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/fastly	480.157s
```
